lib/datadog/grok/src/parse_grok_rules.rs

```diff
@@ -0,0 +1,330 @@
+        )
+    })?;
+    let mut rule_def = split
+        .next()
+        .ok_or_else(|| {
+            Error::InvalidGrokExpression(
+                rule.to_string(),
+                "format must be: 'ruleName definition'".into(),
+            )
+        })?
+        .to_string();
+
+    let rule_def_cloned = rule_def.clone();
+    lazy_static! {
+        static ref GROK_PATTERN_RE: onig::Regex =
+            onig::Regex::new(r#"%\{([^"\}]|(?<!\+)"(\\"|[^"])*(?<!\+)")+\}"#).unwrap();
+    }
+    // find all patterns %{}
+    let raw_grok_patterns = GROK_PATTERN_RE
+        .find_iter(rule_def_cloned.as_str())
+        .map(|(start, end)| &rule_def_cloned[start..end])
+        .collect::<Vec<&str>>();
+    // parse them
+    let mut grok_patterns = raw_grok_patterns
+        .iter()
+        .map(|pattern| {
+            parse_grok_pattern(pattern)
+                .map_err(|e| Error::InvalidGrokExpression(pattern.to_string(), e.to_string()))
+        })
+        .collect::<Result<Vec<GrokPattern>, Error>>()?;
+
+    grok_patterns = index_repeated_fields(grok_patterns);
+
+    let mut filters: HashMap<LookupBuf, Vec<GrokFilter>> = HashMap::new();
+    let pure_grok_patterns: Vec<String> = grok_patterns
+        .iter()
+        .map(|pattern| purify_grok_pattern(&pattern, &mut filters, &mut parsed_rules))
+        .collect::<Result<Vec<String>, Error>>()?;
+
+    // replace grok patterns with "purified" ones
+    for (r, pure) in raw_grok_patterns.iter().zip(pure_grok_patterns.iter()) {
+        rule_def = rule_def.replacen(r, pure.as_str(), 1);
+    }
+
+    // collect all filters to apply later
+    for pattern in grok_patterns {
+        if let GrokPattern {
+            destination:
+                Some(Destination {
+                    filter_fn: Some(ref filter),
+                    ..
+                }),
+            ..
+        } = pattern
+        {
+            let dest = pattern.destination.as_ref().unwrap();
+            let filter = GrokFilter::try_from(filter)?;
+            filters
+                .entry(dest.path.clone())
+                .and_modify(|v| v.push(filter.clone()))
+                .or_insert_with(|| vec![filter.clone()]);
+        }
+    }
+
+    let mut pattern = String::new();
+    pattern.push('^');
+    pattern.push_str(rule_def.as_str());
+    pattern.push('$');
+
+    // store rule definitions and filters in case this rule is referenced in the next rules
+    parsed_rules.insert(
+        rule_name,
+        ParsedGrokRule {
+            definition: rule_def,
+            filters: filters.clone(),
+        },
+    );
+
+    let pattern = Arc::new(
+        grok.compile(&pattern, true)
+            .map_err(|e| Error::InvalidGrokExpression(pattern, e.to_string()))?,
+    );
+
+    Ok(GrokRule { pattern, filters })
+}
+
+/// replaces repeated field names with indexed versions, e.g. : field.name, field.name -> field.name.0, field.name.1 to avoid collisions in grok
```

Comment:
```suggestion
/// Replaces repeated field names with indexed versions, e.g. : field.name, field.name -> field.name.0, field.name.1 to avoid collisions in grok.
```